### PR TITLE
Make environment variables access platform-specific

### LINKED ISSUE #9279: Calling `System.env` and related methods do not work on Scala.js
When using ZIO in a Scala.js project, the `zio.System.env` method and related environment variable access functions do not retrieve environment variables as expected. This issue occurs even when environment variables are correctly set in the Node.js environment using `jsEnv` configurations in `build.sbt`.

### Steps to Reproduce:

1. **Set Up a Scala.js Project with ZIO Dependencies**

   Include ZIO in your `build.sbt`:

   ```scala
   libraryDependencies += "dev.zio" %%% "zio" % "2.1.11"
   ```

2. **Configure `jsEnv` to Pass Environment Variables to Node.js**

   ```scala
   jsEnv := new NodeJSEnv(
     NodeJSEnv.Config().withEnv(Map("FOO" -> "hello", "BAR" -> "world"))
   )
   ```

3. **Attempt to Access Environment Variables Using `zio.System.env`**

   ```scala
   package com.example

   import zio.{ZIO, ZLayer, System}

   case class Config(
     foo: String,
     bar: String
   )

   object Config {
     val live: ZLayer[Any, Exception, Config] = ZLayer(
       for {
         foo <- System.env("FOO").someOrFail(new Exception("FOO not set"))
         bar <- System.env("BAR").someOrFail(new Exception("BAR not set"))
       } yield Config(foo, bar)
     )
   }
   ```

4. **Run the Scala.js Application**

   Execute the application and observe the behavior.

**Expected Behavior:**

- The application should retrieve the `FOO` and `BAR` environment variables.
- No exceptions related to missing environment variables should be thrown.

**Actual Behavior:**

- The application fails to retrieve the environment variables.
- Exceptions are thrown indicating that `FOO` and `BAR` are not set.

### Analysis:

The issue appears because `zio.System.env` relies on the underlying platform's standard library to access environment variables. In Scala.js, `System.getenv` is not available due to JavaScript limitations. Therefore, `zio.System.env` does not function correctly in a Scala.js environment.

**Suggested Solution:**

To address this issue, it would be beneficial for ZIO to provide a Scala.js-specific implementation of `zio.System`, similar to how `ClockPlatformSpecific` is used to handle platform-specific clock functionalities. By creating a `SystemPlatformSpecific` trait for Scala.js, ZIO can directly interact with the Node.js `process.env` object to access environment variables when running under Node.js.

**Proposed Implementation:**

1. **Create a `SystemPlatformSpecific` Trait for Scala.js**

   In the `zio` package, create a new file `SystemPlatformSpecific.scala` specifically for Scala.js:

   ```scala
   // zio/SystemPlatformSpecific.scala (Scala.js version)

   package zio

   import scala.scalajs.js
   import scala.scalajs.js.Dynamic.global

   private[zio] trait SystemPlatformSpecific { self: System.type =>

     def env(variable: => String)(implicit trace: Trace): IO[SecurityException, Option[String]] =
       ZIO.attempt {
         val env = global.process.env.asInstanceOf[js.Dictionary[String]]
         env.get(variable)
       }.refineToOrDie[SecurityException]

     // Implement other methods as needed, such as system properties or line separators
   }
   ```

   This trait provides a Scala.js-specific implementation of `env` by accessing `global.process.env`, which is available when running under Node.js.

2. **Modify the `System` Object to Extend `SystemPlatformSpecific`**

   Ensure that the `System` object extends `SystemPlatformSpecific`, so that the platform-specific implementation is included:

   ```scala
   // zio/System.scala

   package zio

   object System extends SystemPlatformSpecific {
     // Existing code...
   }
   ```

3. **Handle Non-Node.js Environments**

   In cases where Scala.js is not running under Node.js (e.g., in a browser), the implementation should gracefully handle the absence of `process.env`:

   ```scala
   // zio/SystemPlatformSpecific.scala (updated)

   private[zio] trait SystemPlatformSpecific { self: System.type =>

     def env(variable: => String)(implicit trace: Trace): IO[SecurityException, Option[String]] =
       ZIO.attempt {
         if (js.typeOf(global.process) != "undefined" && js.typeOf(global.process.env) != "undefined") {
           val env = global.process.env.asInstanceOf[js.Dictionary[String]]
           env.get(variable)
         } else {
           None
         }
       }.refineToOrDie[SecurityException]
   }
   ```

**Workaround:**

As a temporary workaround, users can define their own method to access environment variables in Scala.js:

```scala
def getEnv(key: String): Option[String] = {
  val env = global.process.env.asInstanceOf[js.Dictionary[String]]
  env.get(variable)
}
```

**Environment:**

- **ZIO Version:** 2.1.11
- **Scala.js Version:** 1.17.0
- **Node.js Version:** 22.11.0
- **Build Tool:** sbt 1.10.5

### PR DESCRIPTION
**Pull request by mhriemers**
Fixes #9279 

/claim #9279

### CODE REVIEW

File: core/js/src/main/scala/zio/SystemPlatformSpecific.scala [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+import java.lang.{System => JSystem}
+import scala.annotation.nowarn
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+import scala.scalajs.js
+import scala.scalajs.js.Dynamic.global
+
+private[zio] trait SystemPlatformSpecific { self: System.type =>
+
+  trait SystemLivePlatformSpecific extends System {
```
Comment (kyri-petrou) [inline core/js/src/main/scala/zio/SystemPlatformSpecific.scala]: Let's make this `private[zio]` as well

File: core/jvm-native/src/main/scala/zio/SystemPlatformSpecific.scala [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import zio.internal.stacktracer.Tracer
+import zio.stacktracer.TracingImplicits.disableAutoTrace
+
+import java.lang.{System => JSystem}
+import scala.annotation.nowarn
+import scala.collection.JavaConverters._
+
+private[zio] trait SystemPlatformSpecific { self: System.type =>
+
+  trait SystemLivePlatformSpecific extends System {
```
Comment (kyri-petrou) [inline core/jvm-native/src/main/scala/zio/SystemPlatformSpecific.scala]: As above

**Review by kyri-petrou — COMMENTED** [review]
Thanks for the contribution! Some really minor comments and I'll merge it in once we resolve an issue with sonatype releases

**Comment by mhriemers** [comment]
@kyri-petrou

I'm considering whether implementing the entire `UnsafeAPI` in `SystemPlatformSpecific` is too broad and whether we should limit the platform-specific implementation to just the `env` and `envs` methods, keeping the rest of the `UnsafeAPI` in the shared code. This might make the implementation more focused and reduce duplication.

Additionally, regarding the system properties and `lineSeparator` implementations in Scala.js, I noticed that these are already handled by the Scala.js library itself:

[Scala.js System.scala implementation](https://github.com/scala-js/scala-js/blob/f6e247e7fab37b035822f99dd0a0584897462190/javalib/src/main/scala/java/lang/System.scala#L188-L283)

```scala
// Excerpt from Scala.js System.scala
def getProperties(): ju.Properties = SystemProperties.getProperties()
def lineSeparator(): String = "\n"
// Other methods...
```

These methods are maintained within the Scala.js standard library, providing compatibility with the Java API. Since they are already implemented and maintained by Scala.js, perhaps we can defer to these existing implementations instead of re-implementing them in our platform-specific code.

What do you think about tailoring the platform-specific code specifically towards environment variables and relying on the existing Scala.js implementations for system properties and line separators? This approach might help us avoid unnecessary duplication and leverage the maintenance work already done by the Scala.js team.

I'm open to your thoughts on this. Do you think it's better to focus the platform-specific code solely on environment variables, or should we keep the complete `UnsafeAPI` implementation within `SystemPlatformSpecific`?

Thanks!

**Comment by mhriemers** [comment]
Something like this:

```diff
diff --git a/core/shared/src/main/scala/zio/System.scala b/core/shared/src/main/scala/zio/System.scala
--- a/core/shared/src/main/scala/zio/System.scala	(revision f092e25d1bfa01eb8dae4dc656ff8e623894ee3b)
+++ b/core/shared/src/main/scala/zio/System.scala	(date 1730986534534)
@@ -101,7 +101,7 @@
     }
 }
 
-object System extends Serializable {
+object System extends SystemPlatformSpecific {
 
   val tag: Tag[System] = Tag[System]
 
@@ -140,7 +140,7 @@
     @transient override val unsafe: UnsafeAPI =
       new UnsafeAPI {
         override def env(variable: String)(implicit unsafe: Unsafe): Option[String] =
-          Option(JSystem.getenv(variable))
+          environmentProvider.env(variable)
 
         override def envOrElse(variable: String, alt: => String)(implicit unsafe: Unsafe): String =
           envOrElseWith(variable, alt)(env)
@@ -148,9 +148,8 @@
         override def envOrOption(variable: String, alt: => Option[String])(implicit unsafe: Unsafe): Option[String] =
           envOrOptionWith(variable, alt)(env)
 
-        @nowarn("msg=JavaConverters")
         override def envs()(implicit unsafe: Unsafe): Map[String, String] =
-          JSystem.getenv.asScala.toMap
+          environmentProvider.envs
 
         override def lineSeparator()(implicit unsafe: Unsafe): String =
           JSystem.lineSeparator
@@ -172,6 +171,11 @@
       }
   }
 
+  private[zio] trait EnvironmentProvider {
+    def env(variable: String): Option[String]
+    def envs: Map[String, String]
+  }
+
   private[zio] def envOrElseWith(variable: String, alt: => String)(env: String => Option[String]): String =
     env(variable).getOrElse(alt)
 

```

**Comment by kyri-petrou** [comment]
@mhriemers yeah I think the approach you proposed makes sense. It should also be bin-compatible so let's go with it 👍

**Comment by mhriemers** [comment]
@kyri-petrou I prepared the commit yesterday, including your review comments.

**Review by kyri-petrou — APPROVED** [review]
LGTM, thanks 🙏

**Pull request merged**